x/sequencer/keeper/rotation.go

```diff
@@ -12,135 +11,82 @@ import (
-	seq, ok := k.GetNextProposer(ctx, rollappId)
-	if ok {
-		return seq
+// ChooseSuccessorForFinishedNotices goes through all sequencers whose notice periods have elapsed.
+// For each proposer, it chooses a successor proposer for their rollapp.
+func (k Keeper) ChooseSuccessorForFinishedNotices(ctx sdk.Context, now time.Time) error {
+	seqs, err := k.NoticeElapsedSequencers(ctx, now)
+	if err != nil {
+		return errorsmod.Wrap(err, "get notice elapsed sequencers")
 	}
-
-	// take the next bonded sequencer to be the proposer. sorted by bond
-	seqs := k.GetSequencersByRollappByStatus(ctx, rollappId, types.Bonded)
-	sort.SliceStable(seqs, func(i, j int) bool {
-		return seqs[i].Tokens.IsAllGT(seqs[j].Tokens)
-	})
-
-	// return the first sequencer that is not the proposer
-	proposer, _ := k.GetProposer(ctx, rollappId)
-	for _, s := range seqs {
-		if s.Address != proposer.Address {
-			return s
+	for _, seq := range seqs {
+		// Successor cannot finish notice. The proposer must finish first and then rotate to the successor.
+		if !k.IsSuccessor(ctx, seq) {
+			k.removeFromNoticeQueue(ctx, seq)
+			if err := k.chooseSuccessor(ctx, seq.RollappId); err != nil {
+				return errorsmod.Wrap(err, "choose successor")
+			}
+			successor := k.GetSuccessor(ctx, seq.RollappId)
+			ctx.EventManager().EmitEvent(
+				sdk.NewEvent(
+					types.EventTypeRotationStarted,
+					sdk.NewAttribute(types.AttributeKeyRollappId, seq.RollappId),
+					sdk.NewAttribute(types.AttributeKeyNextProposer, successor.Address),
+					sdk.NewAttribute(types.AttributeKeyRewardAddr, successor.RewardAddr),
+					sdk.NewAttribute(types.AttributeKeyWhitelistedRelayers, strings.Join(successor.WhitelistedRelayers, ",")),
+				),
+			)
 		}
 	}
-
-	return types.Sequencer{}
+	return nil
 }
 
-// startRotation sets the nextSequencer for the rollapp.
-// This function will not clear the current proposer
-// This function called when the sequencer has finished its notice period
-func (k Keeper) startRotation(ctx sdk.Context, rollappId string) {
-	// next proposer can be empty if there are no bonded sequencers available
-	nextProposer := k.ExpectedNextProposer(ctx, rollappId)
-	k.setNextProposer(ctx, rollappId, nextProposer.Address)
-
-	k.Logger(ctx).Info("rotation started", "rollappId", rollappId, "nextProposer", nextProposer.Address)
-
-	// TODO: use corresponding typed event
-	ctx.EventManager().EmitEvent(
-		sdk.NewEvent(
-			types.EventTypeRotationStarted,
-			sdk.NewAttribute(types.AttributeKeyRollappId, rollappId),
-			sdk.NewAttribute(types.AttributeKeyNextProposer, nextProposer.Address),
-			sdk.NewAttribute(types.AttributeKeyRewardAddr, nextProposer.RewardAddr),
-			sdk.NewAttribute(types.AttributeKeyWhitelistedRelayers, strings.Join(nextProposer.WhitelistedRelayers, ",")),
-		),
-	)
+func (k Keeper) awaitingLastProposerBlock(ctx sdk.Context, rollapp string) bool {
+	proposer := k.GetProposer(ctx, rollapp)
+	return proposer.NoticeElapsed(ctx.BlockTime())
 }
```

Comment:
we discussed in the past that this is not reliable enough.
`awaitingLastProposerBlock` should be true only after `chooseSuccessor` is called

Reply:
Right, good spot, I forgot to move ChooseSuccessorForFinishedNotices to begin block!